include/crow/middlewares/session.h

```diff
@@ -0,0 +1,633 @@
+            } while (store_.contains(id));
+            return id;
+        }
+
+        std::string load_id(const CookieParser::context& cookies)
+        {
+            auto session_raw = cookies.get_cookie(cookie_.name());
+
+            auto pos = session_raw.find(":");
+            if (pos <= 0 || pos >= session_raw.size() - 1) return "";
+
+            auto session_id = session_raw.substr(0, pos);
+            auto session_hmac = session_raw.substr(pos + 1, session_raw.size() - pos - 1);
+
+            return hash_id(session_id) == session_hmac ? session_id : "";
+        }
+
+        void store_id(CookieParser::context& cookies, const std::string& session_id)
+        {
+            cookie_.value(session_id + ":" + hash_id(session_id));
+            cookies.set_cookie(cookie_);
+        }
+
+        std::string hash_id(const std::string& key) const
+        {
+            sha1::SHA1 s;
+            std::string input = key + "/" + secret_key_;
+            s.processBytes(input.data(), input.size());
+            uint8_t digest[20];
+            s.getDigestBytes(digest);
+            return utility::base64encode((unsigned char*)digest, 20);
+        }
+
+    private:
+        std::string secret_key_;
+        int id_length_;
+
+        // prototype for cookie
+        CookieParser::Cookie cookie_;
+
+        Store store_;
+
+        // mutexes are immovable
+        std::unique_ptr<std::mutex> mutex_;
+        std::unordered_map<std::string, std::shared_ptr<session::CachedSession>> cache_;
+    };
+
+    /// InMemoryStore stores all entries in memory
+    struct InMemoryStore
+    {
+        // Load a value into the session cache.
+        // A load is always followed by a save, no loads happen consecutively
+        void load(session::CachedSession& cn)
+        {
+            // load & stores happen sequentially, so moving is safe
+            cn.entries = std::move(entries[cn.session_id]);
+        }
+
+        // Persist session data
+        void save(session::CachedSession& cn)
+        {
+            entries[cn.session_id] = std::move(cn.entries);
+            // cn.dirty is a list of changed keys since the last load
+        }
+
+        bool contains(const std::string& key)
+        {
+            return entries.count(key) > 0;
+        }
+
+        std::unordered_map<std::string, std::unordered_map<std::string, session::multi_value>> entries;
+    };
+
+    // FileStore stores all data as json files in a folder.
+    // Files are deleted after expiration. Expiration refreshes are automatically picked up.
+    struct FileStore
+    {
+        FileStore(const std::string& folder, uint64_t expiration_seconds = /*month*/ 30 * 24 * 60 * 60):
+          path_(folder), expiration_seconds_(expiration_seconds)
+        {
+            std::ifstream ifs(get_filename(".expirations", false));
+
+            auto current_ts = chrono_time();
+            std::string key;
+            uint64_t time;
+            while (ifs >> key >> time)
```

Comment:
I'm not sure I understand this line...

Reply:
[This](https://en.cppreference.com/w/cpp/io/basic_ios/operator_bool) example demonstrates the implicit bool conversion of such streams. `std::ifstream` derives from `std::basic_ios` through its parent.

Reply:
Wow, Thanks a lot @luca-schlecker!